core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -590,10 +590,21 @@ private[spark] class BlockManager(
   private def doGetRemote(blockId: BlockId, asBlockResult: Boolean): Option[Any] = {
     require(blockId != null, "BlockId is null")
     val locations = Random.shuffle(master.getLocations(blockId))
+    var failTimes = 0
     for (loc <- locations) {
       logDebug(s"Getting remote block $blockId from $loc")
-      val data = blockTransferService.fetchBlockSync(
-        loc.host, loc.port, loc.executorId, blockId.toString).nioByteBuffer()
+      val data = try {
+        blockTransferService.fetchBlockSync(
+          loc.host, loc.port, loc.executorId, blockId.toString).nioByteBuffer()
+      } catch {
+        case e: IOException if failTimes < locations.size - 1 =>
+          // Return null when IOException throw, so we can fetch block
+          // from another location if there still have locations
+          failTimes += 1
+          logWarning(s"Try ${failTimes} times getting remote block $blockId from $loc failed:", e)
```

Comment:
I wonder if we should save the last exception, and throw it if we don't find the block in any of the locations.  that would be more consistent with the old behavior (throw an exception), but then again, the api returns an option, so maybe we should just swallow all exceptions and return `None` as you have here.  Also now I see what Sean meant earlier when he suggested pushing this down into `fetchBlockSync`.  It looks as if this method is expecting `fetchBlockSync` to return null when it encounters an error, so maybe we should just change the behavior there after all.  @aarondav any thoughts?  I'm leaning towards pushing it into `fetchBlockSync` -- have it swallow the exception and return null. (this is the only use of `fetchBlockSync`.)